apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,441 @@
+
+	// The status of the post update tasks that are associated with current stage.
+	// +kubebuilder:validation:Optional
+	AfterStageTaskStatus []AfterStageTaskStatus `json:"currentPostStageTaskStatus,omitempty"`
+}
+
+// ClusterUpdateStatus defines the status of the update run on a cluster.
+type ClusterUpdateStatus struct {
+	// The name of the cluster.
+	// +kubebuilder:validation:Required
+	ClusterName string `json:"clusterName"`
+
+	// The status of the update run in the stage.
+	// +kubebuilder:validation:Required
+	Status UpdateStatus `json:"status"`
+}
+
+// UpdateStatus defines the status of a resource in an update run.
+type UpdateStatus struct {
+	// The status of the update on the cluster.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=NotStarted;Running;Completed;Failed;WaitingForApproval
+	State UpdateState `json:"state"`
+
+	// message is a human readable message indicating details about the update status.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxLength=1024
+	Message string `json:"message,omitempty"`
+
+	// The time when the update started on the resource.
+	// +kubebuilder:validation:Optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// The time when the update completed on the resource.
+	// +kubebuilder:validation:Optional
+	CompletedTime *metav1.Time `json:"completedTime,omitempty"`
+
+	// The error detail if the update failed on the resource.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MaxLength=4096
+	ErrorDetail string `json:"errorDetail,omitempty"`
+}
+
+// UpdateState identifies the state of resource in an update run.
+// +enum
+type UpdateState string
+
+const (
+	// UpdateStateNotStarted indicates the update run is not started on this resource.
+	UpdateStateNotStarted UpdateState = "NotStarted"
+
+	// UpdateStateRunning indicates the update run is running on this resource.
+	UpdateStateRunning UpdateState = "Running"
+
+	// UpdateStateCompleted indicates the update run is completed on this resource.
+	UpdateStateCompleted UpdateState = "Completed"
+
+	// UpdateStateFailed indicates the update run is failed on this resource.
+	UpdateStateFailed UpdateState = "Failed"
+
+	// UpdateStateWaitingForApproval indicates the update run is waiting for approval on this resource.
+	UpdateStateWaitingForApproval UpdateState = "WaitingForApproval"
+)
+
+type AfterStageTaskStatus struct {
+	// The type of the post update task.
+	// +kubebuilder:validation:Enum=TimedWait;Approval
+	// +kubebuilder:validation:Required
+	Type AfterStageTaskType `json:"type"`
+
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	//
+	// Conditions is an array of current observed conditions for the specific type of post update task.
+	// Known conditions are "ApprovalRequestCreated", "WaitTimeElapsed", and "ApprovalRequestApproved".
+	// +kubebuilder:validation:Optional
+	Conditions []metav1.Condition `json:"conditions"`
```

Comment:
Can you reuse UpdateStatus instead of Conditions? Maybe rename to StepStatus? Everything: a stage, a cluster, a after stage task is a step in the staged update run.

You also need to record start time and complete time of the step. This is especially useful for TimedWait so the user knows when the wait started and when it ended.

Reply:
The recommended in k8s is to use condition to record information for each state (running, wait, completed). I think it's better to at least keep the overall state in condition instead of a single status. We can convert all the state back into conditions too but it's little bit too much.

Reply:
How about this? We use conditions at the updaterun level, but use updatestatus for stages and below.

Better to have a start time and end time for these AfterStageTasks too. So it is better to use UpdateStatus instead conditions for AfterStageTasks.

Reply:
The condition states (ApprovalRequestCreated, ApprovalRequestApproved, WaitTimeElapsed) in the afterStageTasks are different from the update status so it's hard to reuse it.

Reply:
NotStarted -> (created approval request) -> waiting -> (approve request) -> completed
NotStarted -> (start wait time) -> waiting -> (wait time has passed) -> completed

UpdateRun is a tree of steps: a step can be a run, a stage, a cluster, a after-stage task.

It is better to have the same state model for each step so customers don't need to learn new state/condition model for each step.

That's why we should have the same state for the run and each after-stage task as well.

Reply:
Fleet RP also have the wait time step and use the same state data model.